feat: How to compile and run nearcore (#721)
near/docs: docs/develop/node/compile-and-run-a-node.md

```

### 5. Get data backup

The node is ready to be started. However the first thing to do is to sync up with the netowk. It means your node needs to download all the headers and blocks like other nodes in the network have. This process might be speed up drastically by using backups.

Download and unpack the tar to `~/.near` https://near-protocol-public.s3.ca-central-1.amazonaws.com/backups/testnet/rpc/data.tar

Or run the commands

```bash
$ wget ~/.near/data.tar https://near-protocol-public.s3.ca-central-1.amazonaws.com/backups/testnet/rpc/data.tar
$ tar -xf ~/.near/data.tar
$ rm ~/.near/data.tar
```

### 6. Run a node

```bash
$ ./target/release/neard --home ~/.near run
```

That's all. The node is running you can see log outputs in your console. It will download a bit of missing data since the backup was done but it shouldn't take much time.

## mainnet

### 1. Clone nearcore project from GitHub

You need to clone [`nearcore`](https://github.com/near/nearcore)

```bash
$ git clone https://github.com/near/nearcore
$ git fetch origin --tags
$ cd nearcore
```

> Checkout to the branch you need if not `master` (default). Latest release is recommended. Please check the [releases page on GitHub](https://github.com/near/nearcore/releases). Current latest is `1.19.0`

Reviewer: ```suggestion
Next, checkout the release branch you need (recommended) if you will not be using the default `master` branch. Please check the [releases page on GitHub](https://github.com/near/nearcore/releases) for the latest release. Currently `1.19.0` as of 5/25/21.

For more information on choosing between `master` and latest release branch [ [click here](/docs/develop/node/compile-and-run-a-node#choosing-your-nearcore-version) ].
```

```bash
$ git checkout tags/1.19.0 -b mynode
```

> ### 2. Compile nearcore binary

Reviewer: ```suggestion
### 2. Compile `nearcore` binary
```

> In the `nearcore` folder run the commands:

Reviewer: ```suggestion
In the `nearcore` folder run the following commands:
```

```bash
$ cargo build --release --package neard --bin neard
```

> This will start the compilation process, it will take some time depending on your machine power (e.g. i9 8-core CPU, 32 GB RAM, SSD it takes approximately 25 minutes)

Reviewer: ```suggestion
This will start the compilation process and will take some time depending on your machine's CPU power. _(e.g. i9 8-core CPU, 32 GB RAM, SSD takes approximately 25 minutes)_
```

The binary path is `nearcore/target/release/neard`

### 3. Initialize working directory

In order to work NEAR node requires to have working directory and a couple of configuration files.

> - `config.json` with a lot of parameters which responsive for how the node will work
> - `genesis.json` a file with all the data the network had started with. In contains initial accounts, contracts, access keys and other records which represents the initial state of the blockchain
> - `node_key.json` a file which contains public and private key for the node. Also includes optional `account_id` parameter which is required to run validator node (not covered in this doc)
> - `data/` folder in which NEAR node will write the state.

Reviewer: ```suggestion
- `config.json` - Configuration parameters which are responsive for how the node will work.
- `genesis.json` - A file with all the data the network started with at genesis. This contains initial accounts, contracts, access keys, and other records which represents the initial state of the blockchain.
- `node_key.json` -  A file which contains a public and private key for the node. Also includes an optional `account_id` parameter which is required to run a validator node (not covered in this doc).
- `data/` -  A folder in which a NEAR node will write it's state.
```

Reviewer: Is there not a validator key for a `mainnet` node?

> Initially the required working directory needs to be generated. The command to generate it looks like:

Reviewer: ```suggestion
Generate the initial required working directory by running:
```

```bash
$ ./target/release/neard --home ~/.near init --chain-id mainnet --download
```

> > You can skip `--home` argument if you are fine with default working directory in `~/.near` or pass different path

Reviewer: ```suggestion
> You can skip the `--home` argument if you are fine with the default working directory in `~/.near`. If not, pass your preferred location.
```

> This command will create the required directory structure, will generate `config.json` and `node_key.json`, will download `genesis.json` for mainnet network.

Reviewer: ```suggestion
This command will create the required directory structure by generating a `config.json`, `node_key.json`, and downloads a `genesis.json` for `mainnet`.
```

### 4. Replacing the `config.json`

> Generated `config.json` is missing `boot_nodes` parameter (it is empty) so we will need to replace it by full one. The file can be downloaded:

Reviewer: ```suggestion
The generated `config.json` will be missing a `boot_nodes` parameter (it is empty) so we will need to replace it with a full one. You can do this one of two ways:
```

> - Download `config.json` https://s3-us-west-1.amazonaws.com/build.nearprotocol.com/nearcore-deploy/mainnet/config.json and replace it in your working dir (`~/.near/config.json`)

Reviewer: ```suggestion
1. Download `config.json` [here](https://s3-us-west-1.amazonaws.com/build.nearprotocol.com/nearcore-deploy/mainnet/config.json) and replace it in your working dir (`~/.near/config.json`).
```

> Or run the commands:

Reviewer: ```suggestion
or

2. Run the following commands:
```

```bash
$ rm ~/.near/config.json
$ wget ~/.near/config.json https://s3-us-west-1.amazonaws.com/build.nearprotocol.com/nearcore-deploy/mainnet/config.json
```

### 5. Get data backup

> The node is ready to be started. However the first thing to do is to sync up with the netowk. It means your node needs to download all the headers and blocks like other nodes in the network have. This process might be speed up drastically by using backups.

Reviewer: ```suggestion
The node is ready to be started however the first thing you need to do is to sync up with the network. This means your node needs to download all of the headers and blocks that other nodes on the network have. This process can be sped up drastically by downloading backups one of two ways:
```

> Download and unpack the tar to `~/.near` https://near-protocol-public.s3.ca-central-1.amazonaws.com/backups/mainnet/rpc/data.tar

Reviewer: ```suggestion
1. Download and unpack the [tar file](https://near-protocol-public.s3.ca-central-1.amazonaws.com/backups/mainnet/rpc/data.tar) to `~/.near`.
```

> Or run the commands

Reviewer: ```suggestion
or

2. Run the following commands:
```

```bash
$ wget ~/.near/data.tar https://near-protocol-public.s3.ca-central-1.amazonaws.com/backups/mainnet/rpc/data.tar
$ tar -xf ~/.near/data.tar
$ rm ~/.near/data.tar
```

> ### 6. Run a node

Reviewer: ```suggestion
### 6. Run the node
```

> 

Reviewer: ```suggestion
To start your node simply run the following command:

```

```bash
$ ./target/release/neard --home ~/.near run
```

> That's all. The node is running you can see log outputs in your console. It will download a bit of missing data since the backup was done but it shouldn't take much time.

Reviewer: ```suggestion
That's all. The node is running and you can see log outputs in your console. It will download a bit of missing data since the last backup was performed but it shouldn't take much time.
```
